cpp/src/sort/segmented_sort.cu

```diff
@@ -0,0 +1,153 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+
+#include <cub/device/device_segmented_radix_sort.cuh>
+#include <iterator>
+#include <memory>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace detail {
+
+// returns row index for each element of the row in list column.
+rmm::device_uvector<size_type> get_list_segment_indices(lists_column_view const& lc,
+                                                        rmm::cuda_stream_view stream)
+{
+  auto sliced_child = lc.get_sliced_child(stream);
+  rmm::device_uvector<size_type> segment_ids(sliced_child.size(), stream);
+
+  auto offsets           = lc.offsets().begin<size_type>() + lc.offset();
+  auto offsets_minus_one = thrust::make_transform_iterator(
+    offsets, [offsets] __device__(auto i) { return i - offsets[0] - 1; });
+  auto counting_iter = thrust::make_counting_iterator<size_type>(0);
+  thrust::lower_bound(rmm::exec_policy(stream),
+                      offsets_minus_one + 1,
+                      offsets_minus_one + lc.size() + 1,
+                      counting_iter,
+                      counting_iter + segment_ids.size(),
+                      segment_ids.begin());
+  return std::move(segment_ids);
+}
+
+std::unique_ptr<column> segmented_sorted_order(table_view const& keys,
+                                               std::vector<order> const& column_order,
+                                               std::vector<null_order> const& null_precedence,
+                                               rmm::cuda_stream_view stream,
+                                               rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(std::all_of(keys.begin(),
+                           keys.end(),
+                           [](column_view const& col) { return col.type().id() == type_id::LIST; }),
+               "segmented_sort only supports lists columns");
```

Comment:
Why? I'd think I'd should be able to do a segmented sort of a normal column. In fact, I filed an issue asking for exactly that ages ago: https://github.com/rapidsai/cudf/issues/4603

Reply:
#6541 requested for list columns.  This will be renamed to `sort_lists` and this check will remain.
`segmented_sort` of normal column/table will be exposed as per #4603 specifications.

Reply:
implemented `segmented_sort` as well.